types.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2024 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package raft
+
+// entryID uniquely identifies a raft log entry.
+//
+// Every entry is associated with a leadership term which issued this entry and
+// initially appended it to the log. There can only be one leader at any term,
+// and a leader never issues entries with the same index.
+type entryID struct {
+	term  uint64
+	index uint64
+}
+
+// logAppend is the metadata that describes a validated raft log append.
+//
+// Every log append is done on behalf of a leader at a specific term. This term
+// does not necessarily match entryID.term of the entries being appended, since
+// a leader can send any entries from earlier terms in its own log.
+//
+// TODO(pav-kv): add entries, and helpers for validating the invariants of the append,
```

Comment:
After we add `entries` to this, we can rename `logAppend` to a more generic `logSlice`. For **any** slice of the log, all these invariants are true, and all the fields are useful.

Specifically, the `unstable` log is a `logSlice` in which `prev` is the snapshot entry ID. The in-storage log is the physical counterpart of `logSlice`.

We can then consolidate most of the append arithmetics in `unstable` into this struct. This makes it a nice unit, that we can test properly without needing the boilerplate for setting up `unstable`/`raftLog` etc.

Reply:
Done in this PR.